File: 单调栈/最多能完成排序的块.go
package 单调栈

func maxChunksToSorted0(arr []int) int {
	var stack []int
	for i := 0; i < len(arr); i++ {
		var maxOfThisPiece int
		if len(stack) != 0 && arr[i] < stack[len(stack)-1] {
			maxOfThisPiece = stack[len(stack)-1]
			for len(stack) != 0 && arr[i] < stack[len(stack)-1] {
				stack = stack[:len(stack)-1]
			}
		}
		stack = append(stack, max(maxOfThisPiece, arr[i]))
	}
	return len(stack)
}

//以下思路完全错误
////本题简单的地方在于arr中的每一个元素都不相同，于是可以采用最原始的大顶栈
//func maxChunksToSorted00(arr []int) int {
//	var stack []int
//	for i := 0; i < len(arr); i++ {
//		for len(stack) != 0 && arr[i] < stack[len(stack)-1] {
//			stack = stack[:len(stack)-1]
//		}
//		stack = append(stack, arr[i])
//	}
//	return len(stack)
//}

File: 贪心/_最多能完成排序的块.java
package 贪心;


import java.util.ArrayDeque;

//错误解法
//45 / 52
public class _最多能完成排序的块 {
    public int maxChunksToSorted(int[] arr) {
        int ptr = 0;
        int res = 0;
        while (ptr < arr.length) {
            int next = ptr;
            for (int i = ptr + 1; i < arr.length; i++) {
                if (arr[i] < arr[ptr]) next = i;
            }
            res++;
            ptr = next + 1;
        }
        return res;
    }
}


//贪心 + 栈
//用一个区块中的最大元素来表示这一个块 ， 在从前向后遍历的过程中，出现比我小的，那你归我这个块 ， 比我大 那你自己成块
//错误
class dwde {
    public int maxChunksToSorted(int[] arr) {
        var stack = new ArrayDeque<Integer>();
        for (int i = 0; i < arr.length; i++) {
            if (stack.isEmpty() || arr[i] > stack.peek()) {
                stack.push(arr[i]);
            }
        }
        return stack.size();
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////
//用一个区块中的最大元素来表示这一个块 ,遇到小的要不断将大的块出栈，最后将能代表这个块的最大数字入栈
class dwdwdew{
    public int maxChunksToSorted(int[] arr) {
        var stack = new ArrayDeque<Integer>();
        for (int i = 0; i < arr.length; i++) {
            int maxOfThisBlock = 0;
            if (!stack.isEmpty() && arr[i]<stack.peek()){
                maxOfThisBlock = stack.peek();
                while(!stack.isEmpty() && arr[i]<stack.peek()) {
                    stack.pop();
                }
            }
            stack.push(Math.max(arr[i], maxOfThisBlock));

        }
        return stack.size();
    }
}
